mautic/mautic
# Tprod 296 Authentication refactoring needed for Symfony 5

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [x]
| Automated tests included?              | [x] 
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | Fixes https://mautic.atlassian.net/browse/TPROD-296



#### Description:

I started with this by looking what is the AuthenticationListener used for and found last changes coming from https://github.com/mautic/mautic/pull/9837 so I got an impression it is used for Oauth2 and wrote a test for it. It turns out it's not that related and the AuthenticationListener that needed refactoring is actually used for user login and SSO. Anyway, the test for Oauth2 is useful to have.

The changes needed to get rid of the deprecated ListenerInterface are actually nicely described in https://stackoverflow.com/questions/56438177/deprecated-the-listenerinterface-turn-your-listeners-into-callables-instead and if I'd follow that instead of trying to figure out the usage from git history and the code then it would save a lot of time.

This PR also changes configuration for the "anonymous" security option from `true` to `'lazy'`. That is a new option since Symfony 4.4 described as

> The lazy anonymous mode prevents the session from being started if there is no need for authorization

So I was hoping it will stop creating session cookies on those routes but it behaves the same as before.

#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Test that you can log in to the instance.
3. Test accessing an API endpoint with Oauth 2 by following steps in https://github.com/mautic/mautic/pull/9837
4. If you have that option, test also SSO login.




<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/11057"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
